Add basic context handling and web commands

imurchie opened the pull request:
Passes all the iOS webview tests, not Safari tests.

jlipps commented on lib/commands/context.js:
```diff
@@ -0,0 +1,472 @@
+};
+
+commands.setContext = async function (name, callback, skipReadyCheck) {
+  function alreadyInContext(desired, current) {
```
`function alreadyInContext (desired, current) {`

jlipps commented on lib/commands/context.js:
```diff
@@ -0,0 +1,472 @@
+        await this.remote.selectPage(pageIdKey, skipReadyCheck);
+        this.curContext = idx;
+      } else {
+        if (name === this.curContext) {
```
yeah isn't this our first check above, at the beginning of the function?

jlipps commented on lib/commands/context.js:
```diff
@@ -0,0 +1,472 @@
+      // }.bind(this);
+      // setTimeout(loopClose, 4000);
+      await B.delay(4000);
+      // TODO: figure out how to do the alert stuff
```
yess....

jlipps commented on lib/commands/context.js:
```diff
@@ -0,0 +1,472 @@
+    // this.remote.pageIdKey = parseInt(keyId, 10);
+    newPage = keyId;
+  } else {
+    let dirty = () => {
```
if we just call this one time, let's make it a self-executing anonymous function so that `dirty` is just a bool. this also deserves a comment.

jlipps commented on lib/commands/context.js:
```diff
@@ -0,0 +1,472 @@
+      return !_.isEqual(item(this.contexts), item(pageArray));
+    };
+
+    // If a window gets navigated to an anchor it doesn't always fire a page callback event
```
or maybe just moving the comment up

jlipps commented on lib/commands/context.js:
```diff
@@ -0,0 +1,472 @@
+        // and in the case of iOS 7.1 in a webview (not in Safari)
+        // we can have the url be `about:blank`
+        if (parseFloat(this.iOSSDKVersion) < 7 ||
+            (this.caps.platformVersion === '7.1' && this.caps.app && this.caps.app.toLowerCase() !== 'safari')) {
```
we should be using `this.opts`

sebv replied:
yeah, just following my bad example, we need to change them in all the files in the repo.

imurchie replied:
Yes, I'll table this as a to-do.

jlipps commented on lib/commands/context.js:
```diff
@@ -0,0 +1,472 @@
+extensions.useNewSafari = function () {
+  return parseFloat(this.iosSdkVersion) >= 8.1 &&
+         parseFloat(this.caps.platformVersion) >= 8.1 &&
+         !this.caps.udid &&
```
same here, probably want to do a global find/re

jlipps commented on lib/commands/context.js:
```diff
@@ -0,0 +1,472 @@
+  this.implicitWaitMs = oldImpWait;
+  await this.nativeTap(res.value.ELEMENT);
+  res = await this.findElements('name', 'Address');
+  let addressEl = res.value[res.value.length -1].ELEMENT;
```
space

jlipps commented on lib/commands/context.js:
```diff
@@ -0,0 +1,472 @@
+      } else {
+        throw new Error('Could not navigate to webview; there are none!');
+      }
+    } else {
```
this should probably go up first under an `if (res)` so it can short-circuit and we need fewer indentation levels

jlipps commented on lib/commands/context.js:
```diff
@@ -0,0 +1,472 @@
+        logger.warn("Could not find any webviews yet, refreshing/retrying");
+        if (this.caps.udid || !this.caps.safari) {
+          await B.delay(spinTime);
+          return await spinHandles();
```
don't understand why we would call spinHandles again just after waiting. if that's all we want to do, why not just wait at the top of spinHandles?

sebv replied:
We don't want to wait the first time.

jlipps commented on lib/commands/context.js:
```diff
@@ -0,0 +1,472 @@
+          return await spinHandles();
+        }
+        try {
+          res = await this.findUIElementOrElements('accessibility id', 'ReloadButton', '', false);
```
surprised that this doesn't just return the value directly; i thought we'd moved away from internal functions knowing anything about jsonwp response structure

sebv replied:
it does, can be renamed el.

jlipps commented on lib/commands/context.js:
```diff
@@ -0,0 +1,472 @@
+
+helpers.closeAlertBeforeTest = async function () {
+  let res = await this.uiAutoClient.sendCommand('au.alertIsPresent()');
+  if (res !== null && !_.isUndefined(res) && res.value === true) {
```
same q about value here

jlipps commented on lib/commands/web.js:
```diff
@@ -0,0 +1,193 @@
+
+commands.title = async function () {
+  if (this.isWebContext()) {
+    let res =  await this.executeAtom('title', []);
```
space

jlipps commented on lib/commands/web.js:
```diff
@@ -0,0 +1,193 @@
+    if (!this.returnedFromExecuteAtom[counter] && !this.selectingNewPage) {
+      logger.debug('atom did not return yet, checking to see if ' +
+                   'we are blocked by an alert');
+      // temporarily act like we're not processing a remote command
```
is this comment still relevant? i don't see us setting any processingRemoteCommand variable here

jlipps commented:
ok, done with first pass!

sebv commented on lib/commands/context.js:
```diff
@@ -0,0 +1,472 @@
+  await B.delay(timeout);
+  if (this.useNewSafari()) {
+    await this.typeAndNavToUrl();
+  } else if (parseInt(this.iosSdkVersion, 10) >= 7 && !this.caps.udid && this.caps.safari) {
```
Let's create some `isRealDevice` and `isSafari` functions.

imurchie replied:
There is an `isRealDevice` (Added on line 501 of device.js).

sebv replied:
Then we should try to use them instead of `this.caps.udid` or `this.caps.safari`. I know is not done yet in a lot of places on master.

sebv commented on lib/commands/web.js:
```diff
@@ -0,0 +1,193 @@
+  return res.value;
+};
+
+commands.click = async function (elementId) {
```
This is going to conflict with the gesture one. This probably need to be a `webClick` helper.

sebv commented:
Let's merge it quick so we get both native/web test working from master before Friday.

imurchie commented:
Addressed the issues. Some things are not yet tested since we only use them in Safari. We will go through and get rid of `this.caps` at a later time.

Depends, now, on https://github.com/appium/appium-remote-debugger/pull/4

jlipps commented on lib/commands/context.js:
```diff
@@ -0,0 +1,461 @@
+  }
+};
+
+extensions.typeAndNavToUrl = async function () {
```
this function is so much easier to read now

jlipps commented on lib/commands/context.js:
```diff
@@ -0,0 +1,461 @@
+  this.implicitWaitMs = oldImpWait;
+  try {
+    let res = await this.nativeTap(el.ELEMENT);
+    if (res.status !== 0) {
```
is this another erroneous assumption about receiving a jsonwp response instead of an actual value? can we not get rid of this try/catch and simply allow `nativeTap` to throw if it doesn't work?

jlipps commented on lib/commands/context.js:
```diff
@@ -0,0 +1,461 @@
+      await this.setContext(latestWindow);
+      await this.remote.cancelPageLoad();
+      return;
+    } else {
```
since we have a return here, no need for the `else`. just dedent the following.

jlipps commented on lib/commands/context.js:
```diff
@@ -0,0 +1,461 @@
+      await this.remote.cancelPageLoad();
+      return;
+    } else {
+      if ((Date.now() - start) < 90000) {
```
I think it would be more elegant if there were one try/catch in here and one `return await spinHandles()`

jlipps commented on lib/commands/context.js:
```diff
@@ -0,0 +1,461 @@
+          await B.delay(spinTime);
+          return await spinHandles();
+        }
+      } else {
```
like a previous comment i made, since this is a tiny little block that ends control of this function, why don't we reverse the boolean and put this up top? then we need fewer indented brace blocks :-)

jlipps commented on lib/commands/context.js:
```diff
@@ -0,0 +1,461 @@
+    await this.uiAutoClient.sendCommand('au.dismissAlert()');
+    logger.debug('Alert banished!');
+    return true;
+  } else {
```
following the same style, no need for the `else` here

jlipps commented on lib/commands/context.js:
```diff
@@ -0,0 +1,461 @@
+helpers.stopRemote = async function (closeWindowBeforeDisconnecting = false) {
+  if (!this.remote) {
+    logger.errorAndThrow('Tried to leave a web frame but were not in one');
+  } else {
```
same here

jlipps commented on lib/commands/web.js:
```diff
@@ -0,0 +1,201 @@
+    this.processingRemoteCmd = true;
+    await this.remote.navToUrl(url);
+    this.processingRemoteCmd = false;
+  } else {
```
same here and in all following. would prefer this to read:

```js
if (!this.isWebContext()) {
  throw new errors.NotImplementedError();
}
// rest of code goes here
```

jlipps commented on lib/commands/web.js:
```diff
@@ -0,0 +1,201 @@
+commands.title = async function () {
+  if (this.isWebContext()) {
+    this.processingRemoteCmd = true;
+    let title =  await this.executeAtom('title', []);
```
doesn't executeAtom already set the `processingRemoteCmd`?

imurchie commented:
Alrighty! Thanks for the comments. It looks much better now.

jlipps commented on lib/commands/context.js:
```diff
@@ -0,0 +1,462 @@
+        throw new errors.NoSuchWindowError();
+      }
+    }
+  } else {
```
hate to nitpick, but this is popping out at me as needing to go up top as well :-) just seems natural to have your chain of "if this isn't the case, then throw an error" up top, with the meat below

imurchie replied:
No need to hate nitpicking.

jlipps commented on lib/commands/context.js:
```diff
@@ -0,0 +1,462 @@
+    if (this.isRealDevice()) {
+      this.remote = new WebKitRemoteDebugger();
+      return this.remote.pageArrayFromJson();
+    } else {
```
because of the return above, don't need an else block here

jlipps commented on lib/commands/context.js:
```diff
@@ -0,0 +1,462 @@
+  let present = await this.uiAutoClient.sendCommand('au.alertIsPresent()');
+  if (!present) {
+    return false;
+  } else {
```
no need for an else with the return

jlipps commented on lib/commands/web.js:
```diff
@@ -0,0 +1,189 @@
+    // in the future, detect whether we have a UIWebView that we can use to
+    // make sense of this command. For now, and otherwise, it's a no-op
+    throw new errors.NotImplementedError();
+  } else {
```
no need for else with throw

jlipps replied:
and following (that's what i had in mind originally, and the benefit---one less level of indentation! one less pair of braces!)

jlipps commented on lib/commands/web.js:
```diff
@@ -0,0 +1,189 @@
+  if (this.isWebContext()) {
+    let atomsElement = this.useAtomsElement(elementId);
+    return await this.executeAtom('click', [atomsElement]);
+  } else {
```
basically just look anywhere you have an `else` in the whole project and chances are you can remove it :-)
